### ISSUE TITLE
Doze mode will discard the service inbetween songs

### ISSUE THREAD

**Issue by AndrewBloom** [android/uamp#349]
This code doesn't show best practice in my opinion:

My error was to copy this behaviour from your sample UAMP. https://github.com/android/uamp/blob/master/common/src/main/java/com/example/android/uamp/media/MusicService.kt

when (updatedState) {
                PlaybackStateCompat.STATE_BUFFERING,
                PlaybackStateCompat.STATE_PLAYING -> {
                    becomingNoisyReceiver.register()

                    /**
                     * This may look strange, but the documentation for [Service.startForeground]
                     * notes that "calling this method does *not* put the service in the started
                     * state itself, even though the name sounds like it."
                     */
                    if (notification != null) {
                        notificationManager.notify(NOW_PLAYING_NOTIFICATION, notification)

                        if (!isForegroundService) {
                            ContextCompat.startForegroundService(
                                applicationContext,
                                Intent(applicationContext, this@MusicService.javaClass)
                            )
                            startForeground(NOW_PLAYING_NOTIFICATION, notification)
                            isForegroundService = true
                        }
                    }
                }
                else -> {
                    becomingNoisyReceiver.unregister()

                    if (isForegroundService) {
                        stopForeground(false)
                        isForegroundService = false

                        // If playback has ended, also stop the service.
                        if (updatedState == PlaybackStateCompat.STATE_NONE) {
                            stopSelf()
                        }

                        if (notification != null) {
                            notificationManager.notify(NOW_PLAYING_NOTIFICATION, notification)
                        } else {
                            removeNowPlayingNotification()
                        }
                    }
                }
            }

This triggers a sequence of stopForeground, startForeground on the switch between songs. This is not important if the device is in standard mode, but in doze mode as soon the service calls stopForeground nothing else is executed, and in my case the playlist manager didn't trigger the next song to be played. So the ActivityManager after ~40sec discards the service. 

This was experienced in my app but i bet yours behave the same

To trigger doze mode:

adb shell dumpsys deviceidle force-idle; adb shell dumpsys battery unplug; adb shell am set-inactive com.yourapp.package true

**Comment by nic0lette** [comment]
Hi,
I'm not able to reproduce this using the sample code itself (that is, the sample never transitions through any state that causes `MusicService` to call `stopForeground`). What state is it going through in your app?

**Comment by AndrewBloom** [comment]
Hi you're right. Your uamp doesn't transition to a paused or stopped state between songs of a playlist. I suppose this is the behaviour of exoplayer. I have a custom player that sends an end of track event that I made it trigger a pause event. So I got that wrong. Apologies.
Regards,Andrea

Sent from Yahoo Mail on Android 
 
  On Fri, 1 May 2020 at 18:54, Lady Nicole<notifications@github.com> wrote:   


Hi,
I'm not able to reproduce this using the sample code itself (that is, the sample never transitions through any state that causes MusicService to call stopForeground). What state is it going through in your app?
